run BATS tests in Cirrus

edsantiago opened the pull request:
I'm running the BATS tests manually once in a while, and
catching several problems each week that make it past
the rest of CI. Since the BATS tests run at RPM gating
time, we need to catch problems earlier. Try running
the tests from Cirrus.

Signed-off-by: Ed Santiago <santiago@redhat.com>

edsantiago commented:
This PR is, at this moment, expected to fail CI due to #2899. I will need to rebase and re-test once that bug is fixed, but am submitting now to make sure that my addition works and causes Cirrus to fail.

mheon commented on contrib/cirrus/integration_test.sh:
```diff
@@ -63,6 +63,6 @@ else
         *) bad_os_id_ver ;;
     esac
-    make localintegration
+    make localintegration localsystem
```
Can we get this in a separate VM? Would require more intervention from Chris, but our full integration tests are already taking 40-ish minutes.

edsantiago replied:
Can we defer optimization? This test takes 2m40s (5 if rootless is tested serially), and we're wasting a lot more time than that with bugs that current CI is not catching.

cevich replied:
FWIW- I wouldn't bother with a separate VM, those take a full minute just to come up.  But running `make localsystem` would be good to do as a separate "systemtest_script" (in `cirrus.yml`), for no other reason than to separate the output.

edsantiago replied:
Is this a duplicate or stale comment? The one-step 'make localintegration localsystem' went away weeks ago; I'm confused by this.

cevich replied:
You're right, under 'files' tab it's correct :confused:   What tha heck github...OIC, it does say 'outdated' at the top.   I think if I hit the 'resolved' button, this thread will go away, lessee...

edsantiago commented:
Guess this needs a lot more work: 'jq' is not installed on the test systems; Ubuntu has a too-ancient version of `timeout`; and I think I must be misunderstanding the `make install` because I'm seeing failures I wouldn't expect in newer podman. Back to the drawing board.

edsantiago commented:
CI failure is in ginkgo, in code I never touched. I can't figure out how to troubleshoot it.

mheon commented:
Your tests are green

edsantiago commented:
@cevich could you PTAL at your convenience, and make sure this fits in with the CI framework? TIA.

cevich commented:
/approve

edsantiago commented:
/hold

edsantiago commented:
This is in limbo until #3217 gets merged; once that happens, I need to rewrite this one and clean it up.

rhatdan commented:
#3217  is merged.

edsantiago commented:
Yes... and I rebased, cleaned up, and resubmitted this one. Am waiting for CI flakes to clear up (about 6 retries so far) before I cancel hold. But it's ready for review, though.

edsantiago commented:
@cevich any idea what's going on here? I've hit the re-test button about ten times, no luck. (Other jobs failed this morning but re-test got them passing).

cevich commented:
@edsantiago I need to catch up again with Matt and Brent this morning.  Master has been broken for several days, and I think they guys are still debugging, but getting close.  It's related to the `mountopts` thing, everybody is affected :confused:

edsantiago commented:
Our failure to run BATS tests in CI is now causing RHEL8 gating tests to fail. I would really, really like to make this PR a priority.

mheon commented:
Needs a rebase to fix CI

mheon commented:
/approve
Code changes LGTM.

edsantiago commented:
I can rebase, but isn't the `mountopt` error still present?

edsantiago commented:
`runc pause` is failing in the non-remote test; the remote test is failing with other errors. None of those look related to my changes. I'm giving up for the week now.

rhatdan commented:
@edsantiago With the mountopt patch, does this pass now?

edsantiago commented:
@rhatdan  it doesn't seem to be a mountopt problem. Latest failure is:
```
Error: invalid configuration, cannot set resources with rootless containers not using cgroups v2 unified mode
```
...plus, sigh, the latest `container restore` PR broke `--help`. I'll need to submit a patch for that. Monday.

rhatdan commented:
The resource tests should not be running rootless.

edsantiago commented:
@cevich tests continue to fail. Latest results are, as you requested, with `-nodes 3` removed. Timeout is gone, but there are still failures; I'm unable to glean anything useful from them. Is it possible that my BATS tests are leaving the VM in a weird state?

cevich commented:
> @cevich tests continue to fail. Latest results are, as you requested, with -nodes 3 removed. Timeout is gone, but there are still failures; I'm unable to glean anything useful from them. Is it possible that my BATS tests are leaving the VM in a weird state?

Okay maybe that's good, last time we had a hanging test, it would NOT reproduce when `-nodes 3` was removed.  Hopefully that's not the case here, looking...

edsantiago commented:
I think the Panic is an actual bug in the test suite; filed #3309 
No idea about the rest.

Also, if you use my GM script, run an update. I've added stuff to make output much more readable.

cevich commented:
There are some "active" flakes still on master, peeping at the last few, they don't match with this :S  I'd guess that issue you filed is a likely candidate worth scrutiny.  Most of what we find as flakes are race-conditions of one kind or another.  Those are almost always sensitive to timing, so the smallest little change made in the PR could be perpetrating that.

Does that same port test fail repeatedly when you re-run the tests?

Looking at the logs deeper...

cevich commented:
Can't do greasemonkey in chrome :disappointed:   For debugging these ginkgo tests, I just search for the first `Failure [` match and work on that. Nearly always the rest of the test failures can either be ignored or were caused (somehow) by the first one)

edsantiago commented:
One of the two previous failures passed on re-test. Other failure is now:
```
[+0260s] Running: podman [options] container checkpoint d06c66dad86a228bd6bcaff342d2a7f8760f243e7f411f75de59a3f7f4b572fe
[+0260s] Error: checkpointing a container requires root
```
This test is in `test/e2e/checkpoint_test.go`, which shows:
```go
    BeforeEach(func() {
        SkipIfRootless()
```
Again, there has to be something obvious I'm missing.

cevich commented:
> Again, there has to be something obvious I'm missing.

Don't be harsh on yourself, these tests are notoriously difficult to debug.  They demand fairly deep knowledge of both the test and sometimes podman internals.   Flakes don't help this one bit, since few people are exposed to the entirety of failures on Master.

I think the checkpoint thing is a flake, peeking now...

mheon commented:
I haven't seen checkpoint flakes before, but we did just merge a rather gigantic checkpoint/restore patch set, so it makes sense.

cevich commented:
Something else odd is happening here as well:

```
[+0385s] Ginkgo ran 1 suite in 6m8.589186099s
[+0385s] Test Suite Failed
[+0385s] make: *** [Makefile:210: ginkgo] Error 1
[16:51:05] END - [+0385s] total duration since START

Exit status: 2
```

That run should have taken twice as long, and concluded with timing info and a summary like this: https://cirrus-ci.com/task/4899535604678656?command=integration_test#L5719

cevich commented:
@mheon no I think Ed's right as well, we should not be running checkpoint tests as non-root.  The `SkipIfRootless()` condition checks eUID is non-zero, which seems correct.  However, I wonder is it appropriate to call `ginkgo.Skip()` within a `BeforeEach()` closure?

cevich commented:
nvm, found an example of this.  It should look similar to:

```
[+0457s] S [SKIPPING] in Spec Setup (BeforeEach) [0.001 seconds]
[+0457s] Podman stats
[+0457s] /var/tmp/go/src/github.com/containers/libpod/test/e2e/stats_test.go:14
[+0457s]   podman stats with bogus container [BeforeEach]
[+0457s]   /var/tmp/go/src/github.com/containers/libpod/test/e2e/stats_test.go:39
[+0457s] 
[+0457s]   This function is not enabled for rootless podman
```

There are indeed some really odd gremlins playing around here :confounded:

edsantiago commented:
Note that UID is definitely nonzero:
```
Hello, my name is some3271dude and I live in /var/tmp/...
...
DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/22922/bus
...
PODMAN_VARLINK_ADDRESS=unix:/tmp/podman-22922
```
I'm really stumped.

cevich commented on Makefile:
```diff
@@ -207,7 +207,7 @@ localunit: test/goecho/goecho varlink_generate
 
 ginkgo:
-	ginkgo -v -tags "$(BUILDTAGS)" $(GINKGOTIMEOUT) -cover -flakeAttempts 3 -progress -trace -noColor -nodes 3 -debug test/e2e/.
+	ginkgo -v -tags "$(BUILDTAGS)" $(GINKGOTIMEOUT) -cover -flakeAttempts 3 -progress -trace -noColor -debug test/e2e/.
```
Friendly reminder to put back `-nodes 3` when appropriate (I'm not sure we're past the odd hanging condition yet or not).

edsantiago replied:
That's why it's in a standalone commit.

cevich replied:
kk gotcha.

cevich commented:
Let's re-run again and see what happens, maybe it goes green and we can pawn the problem off onto some other poor fool (which will probably end up being me, as my luck goes for these things) :grinning: 

Ref: Oddball rootless log: https://api.cirrus-ci.com/v1/task/6435038086823936/logs/integration_test.log

cevich commented:
@mheon Noting it's checking e-UID and not simple UID.  For the rootless tests, remember we `ssh user@localhost /path/to/script.sh`.  IIRC, ssh forks/execs the handling process as the user, no?  Is there some way e-UID could not == UID messing up the skip?  We're missing something.

cevich commented:
Mmm, nope, euid is def. the user's in the process run by the sshd fork :confused: 

Was there a solar-flare...Can we blame this on EMI or "dark energy" somehow?

cevich commented on .cirrus.yml:
```diff
@@ -315,6 +316,7 @@ special_testing_rootless_task:
 
     setup_environment_script: '$SCRIPT_BASE/setup_environment.sh |& ${TIMESTAMP}'
+    system_test_script: '$SCRIPT_BASE/system_test.sh |& ${TIMESTAMP}'
     integration_test_script: '$SCRIPT_BASE/integration_test.sh |& ${TIMESTAMP}'
```
(Here and elsewhere) It's hurting my brain that we're running system tests before integration tests.

Is it reasonable to think, a new system-test, could perturb the environment such that it passes, but leads to a future integration-test failure for a non-obvious reason?

In other words, the int. tests are already difficult to debug, if we push folks toward writing new system-tests, could that be shooting our integration-tests in the foot?

---even if not---

It might be worth reversing the order temporarily to see if the oddball integration-test flakes go away :smile:

edsantiago replied:
I did that because I thought a 3-minute test would be nice to run before a 30-minute one; quicker notice of failure and all that. But I'll change it and see.

cevich replied:
Ya I suspected as much.  The last run failed in the same way, same test.  Let's see if it does that a third time now with the order reversed...

edsantiago commented:
Well, huh.

cevich commented:
> Well, huh.

double-huh.

cevich commented:
Checking to see if any of these are friends of mine we can safely re-run...

edsantiago commented:
Don't bother. It's a mess. Basicalluy, two PRs snuck in that broke my tests. Trying to fix them now.

edsantiago commented:
(Those don't explain the integration_test failures, but I have to force a new push anyway, so IO'm just going to ignore those)

cevich commented:
...but there are compile warnings now...is something changing the code between system and integration tests? Like a 'git reset' or somesuch?  Okay maybe not,  those warnings are on master too.

I'm still struck by how different the integration tests "look", maybe it's the missing `-nodes 3`?  It might be time to get dirty-hands with `hack/get_ci_vm.sh` on this PR.  We've found a number of nasty things before, by looking while the tests run.

I'll give it a try once you rebase and push.

edsantiago commented:
The `-nodes 3` is back; I removed it as part of my last rebase.

Unfortunately master is currently broken (#3325 and one more that I'm filing) so there's no point in proceeding with this PR until those get fixed.

edsantiago commented:
This makes no sense. System test should be failing: one failure as root, three as nonroot. (They're running. I checked the logs.) This has been a very surreal week.

cevich commented:
Welp, tests look green and changes LGTM.  Plus tomorrow's Friday, which is traditionally when everything goes FUBAR anyway, so :wine_glass: and deal with things Monday?

rhatdan commented:
Looks like the tests passed?

edsantiago commented:
@rhatdan yes, and I think it's because we have no f30 testing hosts. As soon as we do, #3325 and #3326 will bite us on nonroot.

But... those are getting good attention, and it's possible they may get fixed soon, so I'm tentatively OK with merging this one in now; but not without @cevich's and @mheon's blessings.

cevich commented:
@edsantiago now you see why some of my PRs tend to grow so large and take so long :smile: 

I'm conflicted on merging this.  On one hand a functional `make localsystem` will benefit development efforts beyond/outside of CI.  On the other, we're sorely behind the curve and late going live with F30 testing.  Speaking of which...I see yet another rebase is needed over there...

What if "activating" the system-tests in CI was commented out temporarily, then the comment-removed when the issues are fixed?    Would that help get this merged or just make the overall situation more complicated?

edsantiago commented:
Rebased against master, which includes #3328. CI passing (other than an iptables flake which passed on rerun) and I expect them to pass once we get f30 images in CI. I'd love to get this merged!

mheon commented:
LGTM
@baude @rhatdan @haircommander @jwhonce @vrothberg PTAL

rhatdan commented:
/lgtm

edsantiago commented:
/hold cancel

edsantiago commented:
Rebased against master, fixed (simple) conflicts. Tests green again. Could I ask for one more lgtm please?

haircommander commented:
/lgtm
